File: apps/web/repositories/api/scene.repository.ts
// API Scene Repository implementation
import type { ISceneRepository } from "../types";
import type {
  SceneOut,
  SceneCreate,
  SceneUpdate,
  SceneActivateResult,
} from "@/types/api";
import { apiClient } from "@/services/api-client";

export class ApiSceneRepository implements ISceneRepository {
  public async list(homeId?: string): Promise<SceneOut[]> {
    return apiClient.get("/scenes", {
      params: homeId ? { home_id: homeId } : undefined,
    });
  }

  public async create(req: SceneCreate): Promise<SceneOut> {
    return apiClient.post("/scenes", req);
  }

  public async get(id: string): Promise<SceneOut> {
    return apiClient.get(`/scenes/${id}`);
  }

  public async update(id: string, req: SceneUpdate): Promise<SceneOut> {
    return apiClient.patch(`/scenes/${id}`, req);
  }

  public async delete(id: string): Promise<void> {
    return apiClient.delete(`/scenes/${id}`);
  }

  public async activate(id: string): Promise<SceneActivateResult> {
    return apiClient.post(`/scenes/${id}/activate`);
  }
}

File: apps/android/core/src/main/java/com/atech/core/data/SceneRepository.kt
package com.atech.core.data

import com.atech.core.common.DomusResult
import com.atech.core.model.Scene
import com.atech.core.model.SceneActivateResult
import com.atech.core.model.SceneCreate
import com.atech.core.model.SceneUpdate
import com.atech.core.network.DomusHttp
import com.atech.core.network.safeApiCall
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.request.patch
import io.ktor.client.request.post
import io.ktor.client.request.setBody

/** Scenes. Mirrors `backend/scenes/router.py`. */
class SceneRepository(private val http: DomusHttp) {
    private val base get() = "${http.config.apiBase}/scenes"

    suspend fun list(homeId: String? = null): DomusResult<List<Scene>> =
        safeApiCall { http.client.get(base) { parameter("home_id", homeId) } }

    suspend fun get(sceneId: String): DomusResult<Scene> =
        safeApiCall { http.client.get("$base/$sceneId") }

    suspend fun create(data: SceneCreate): DomusResult<Scene> =
        safeApiCall { http.client.post(base) { setBody(data) } }

    suspend fun update(sceneId: String, data: SceneUpdate): DomusResult<Scene> =
        safeApiCall { http.client.patch("$base/$sceneId") { setBody(data) } }

    suspend fun delete(sceneId: String): DomusResult<Unit> =
        safeApiCall { http.client.delete("$base/$sceneId") }

    suspend fun activate(sceneId: String): DomusResult<SceneActivateResult> =
        safeApiCall { http.client.post("$base/$sceneId/activate") }
}
